File: cpp/Service/UserService.cpp
#include "UserService.h"

int UserService::CreateNewUser(User newUser)
{
    string username = newUser.GetUsername();

    User* existingUser = this->m_repository->GetUser(username);
    if (existingUser != NULL)
    {
        return -1;
    }

    this->m_repository->Add(&newUser);

    return 0;
}

void UserService::SetUserRepository(UserRepository* repository)
{
    this->m_repository = repository;
}

File: ruby/service/user_service.rb
require_relative "../model/user.rb"
require_relative "../repository/user_repository.rb"


class DuplicateUserException < StandardError; end


class UserService

    def initialize(user_repository)
        @user_repository = user_repository
    end

    def create_user(user)
        raise DuplicateUserException, "#{user.username} already exists in our system" if @user_repository.get(user.username)

        @user_repository.add(user)
    end
end